### ISSUE TITLE
TK2211-0111. Accessibility

### ISSUE THREAD

**Issue by JohnXLivingston** [JohnXLivingston/peertube-plugin-livechat#118]
With the help of an external audit, this task aims to make sure that the livechat plugin respects accessibility standards. This task may include some work on the upstream ConverseJS project.

Here are the issues revealed by the audit.
It was divided in two parts: the livechat plugin itself, and the documentation website.

## livechat plugin

|Ok|Nr|Issue|Component|Action|
|--|--|--|--|--|
| [x] | 1 | user can navigate out of some modals | Converse | Fixed thanks to bootstrap 5 integration in Converse|
| [x] | 2 | buttons in some modals are identified as 'blank', 'clickable' or not narrated at all by screen readers | Converse and livechat | Fixed by adding type="button" or role="button", on both Converse and livechat |
| [x] | 3 | 'enter' or 'space' closes some modals instead of calling the expected function | Converse | I think this was related to missing role="button" and aria-label attribute on modal close button. This should be fixed thanks to bootstrap 5 integration in Converse. |
| [x] | 4 | When NVDA screen reader is enabled, there are some fields that can't be filled | livechat | Seems that it is resolved with bootstrap v5 integration (i don't have windows to test, but someone else tested for me) |
| [x] | 5 | Modal headers do not appear in the elements list | Converse | Should be fixed thanks to bootstrap 5 integration in Converse |
| [x] | 6 | Screen readers are unable to read text in some modals |Converse | Fixed thanks to bootstrap 5 integration in Converse|
| [x] | 7 | The live chat's ‘message’ box does not allow navigation out of it using the TAB key, only permitting exit by using arrow keys after manually disabling the input field with INSERT+SPACEBAR. Because of this, the toolbar unterneath the video, with the like/dislike button, is also difficult to reach with the keyboard. | Converse |It is because of the Converse Autocomplete class. I will propose a fix: shift+tab should never be cancelled, and tab should not be cancelled if message is empty|
| [x] | 8 | Lacks of a programmatically determinable heading to easily identify and interact with the chat component| livechat |Added a role="region" and an aria-label="Chat" on the chat container  |
| [x] | 9 | Contrast ratio of the menu button |Converse |Seems fixed since bootstrap 5 integration. |
| [x] | 10 | The ‘menu’ button above the chat lacks discernible text, which can lead to confusion and inaccessibility for users relying on screen readers. | Converse | Fixed by adding a aria-label and hidding the icon |
| [x] | 11 | When opening the chat in a new window, the title indicating the ‘live’ for which the chat is used is not formatted as an HTML header. This prevents it from being navigated towards directly. | Converse | Added role="heading" and aria-level="2". Added other similar things in Converse, to fix similar issues that are not visible in livechat |
| - | 12 | In the ‘Share chat link’ modal, when users attempt to copy the URL with the provided  button, a notification box appears indicating that the action was successful. However, the notification is not announced by screen readers, and focus is shifted to the link itself. Not reading out messages prevents screen reader users from receiving important feedback that confirms their actions, potentially causing confusion and diminishing the user experience.| Peertube | Will be fixed by the Peertube team |
| [ ] | 13 | When accessing the ‘online help’ page through the button above the chat, users find images that lack sufficient alternative (alt) text |
| [x] | 14 | The ‘hide occupants’ button, which appears after clicking the ‘show participants’ button, lacks discernible text and is not implemented as a native HTML ‘button’ element.| livechat | Fixed |
| [x] | 15 | The profile pictures next to chat messages  of members are not marked as decorative.| Converse | Fixed |
| [x] | 16 | Next to the password label, there is an input field with the class ‘hidden-username’ that is not visible to users but is still read by screen readers.| Converse | Fixed with aria-hidden="true" |
| [x] | 17 | Next to each chat message, there is a ‘chat message actions’ dropdown that appears only when hovered over with a mouse (and when one is the moderator of the chat). This button does not become visible when screen reader focus is moved to it via keyboard navigation. | Converse | Fixed |
| - | 18 | When navigating the buttons underneath the video, it is difficult to see the focus on the toolbar with the like and dislike button. As well as the little icon behind the comments. A focus should be visible here in order to help users understand where they are on the page. | Peertube | Will be fixed by the Peertube team |
| [x] | 19 | Under the ‘Configure your live's chatrooms’ section under ‘Chatrooms’, each chatroom avatar lacks discernible text. Users who rely on screen readers can't identify the purpose  or content associated with the avatar because of this.| livechat |Fixed by adding a `aria-hidden="true"` on the avatar|
| [x] | 20 | In the section labeled ‘moderation bot options’ under ‘channel options’, the form elements under the some headers are missing labels| livechat | Fixed. Added title attributes on inputs. |
| - | 21 |  When configuration options in ‘Channel Options’ are saved, an alert notifies the userConfiguration of live's of successful saving. However, the screen reader fails to read the alert's content, instead announcing only ‘alert button close-button unavailable’. This issue also occurs when saving configuration options in ‘Channel emojis’.|Peertube|Will be fixed by the Peertube team|
| [x] | 22 | in the section ‘channel emojis’, the form  elements that show up when adding a new emoji are missing labels. |livechat|Fixed. Added title attributes on inputs.|
| - | 23 | The ‘browse’ option when trying to add a  new channel emoji in the section ‘channel emojis’ has insufficient contrast.|Peertube dark theme|Will be fixed by the Peertube team|
| - | 24 |  Across the entire website with dark mode enabled, all buttons (and other elements)featuring an orange background (#f2690d) and white text or symbols (#ffffff) fail to meet the minimum contrast ratio required for accessibility. |Peertube dark theme|Will be fixed by the Peertube team|

## documentation

|Ok|Nr|Issue|Action|
|--|--|--|--|
| [x] | 1 | The contrast between the foreground and background of several elements is not sufficient | Fixed |
| [x] | 2 | Content that appears in the same way on most pages, such as the main navigation, should have a bypass link (skiplink). |This seems fixed after migration from hugo-theme-learn to hugo-theme-relearn. With this theme, the left sidebar is after the main content in the DOM, so left menu entries come after links in the main content.|
| [x] | 3 | Resizing the text on the page/app to 200% causes the breadcrumbs to be fall of the screen.  | Fixed, now breadcrumbs will wrap.|
| [x] | 4 |  Links must be distinguishable without relying on color. The hyperlinks Grav and Hugo on the left-hand side at the bottom of the navigation menu are only recognizable as links through color. | Fixed.|
| [x] | 5 |  You can’t reach several elements when using the keyboard. | Fixed after migration from hugo-theme-learn to hugo-theme-relearn. |
| [x] | 6 | Links must have discernible text. Several links on the site do not have discernible text. | Fixed |
| [x] | 7 | UL and OL elements must only directly contain LI, SCRIPT or TEMPLATE. The lists on several pages (for  example: Documentation), use elements directly within UL elements.  |  Fixed after migration from hugo-theme-learn to hugo-theme-relearn. |
| [x] | 8 | Select element must have an accessible name. The select element  select a language in the bottom of the navigation bar does not have an accessible name. | Fixed after migration from hugo-theme-learn to hugo-theme-relearn. |
| [x] | 9 | The interactive element to select a language doesn’t have a visible focus (are highlighted) when it receives focus using the keyboard. | Fixed after migration from hugo-theme-learn to hugo-theme-relearn. |
| [x] | 10 | When using keyboard only (tabbing through the actionable items), the keyboard focus indicator does not have enough contrast (should be at least 3:1). When selecting a link on the page, a double underline is used as the keyboard focus indicator instead of the usual rectangle around the link. Because the contrast of the links on these pages is not sufficient the contrast of the keyboard focus indicator is also not sufficient in this case. | Fixed (adding a border on focus. |
| [ ] | 11 | Since it is possible to change the language of the website, changing the language then also changes the LANG attribute of the HTML element. However, because large parts of the pages are not yet translated these are still displayed in English. It is not programmatically indicated which parts are still in English and which are in the chosen language. | Created an [issue](https://github.com/mquinson/po4a/issues/516) for po4a |
| [x] | 12 | The ‘warning’, ‘tip’ and ‘info’ boxes do not display these titles as text making them unreadable for a screen reader. |Fixed after migration from hugo-theme-learn to hugo-theme-relearn. The label is now read. |
| [ ] | 13 | The role of the ‘warning’, ‘tip’ and ‘info’ boxes is not communicated trough another mechanism other than visual. Because these boxes do convey information, they should have a title.  | Created an [issue](https://github.com/McShelby/hugo-theme-relearn/issues/897) for hugo-theme-relearn. This issue is not easy to fix. I had a long discussion with an a11y expert, and did not find the solution yet.|

Other finding:

|Ok|Nr|Issue|Action|
|--|--|--|--|
| [ ] | 1 | While the images are still readable at 200% zoom, they don’t scale well. | Don't know how to fix. Should find best practices for screenshots.|
| [x] | 2 | While showing the focus of links using a double underline instead of a single is technically correct, it is less clear than showing a full box around the focused link. | Fixed. |
